k9mail/src/main/java/com/fsck/k9/service/MailListViewWidgetService.java

```diff
@@ -0,0 +1,194 @@
+package com.fsck.k9.service;
+
+import android.appwidget.AppWidgetManager;
+import android.content.Context;
+import android.content.Intent;
+import android.database.Cursor;
+import android.graphics.Color;
+import android.media.Image;
+import android.os.Binder;
+import android.view.View;
+import android.widget.RemoteViews;
+import android.widget.RemoteViewsService;
+import android.widget.TextView;
+import com.fsck.k9.R;
+import com.fsck.k9.provider.MessageProvider;
+
+import java.util.ArrayList;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.Locale;
+
+/**
+ *
+ */
+public class MailListViewWidgetService extends RemoteViewsService {
+
+    @Override
+    public RemoteViewsFactory onGetViewFactory(Intent intent) {
+        return new MailListRemoteViewFactory(this.getApplicationContext(), intent);
+    }
+}
+
+class MailListRemoteViewFactory implements RemoteViewsService.RemoteViewsFactory {
+    private static String [] MAIL_LIST_PROJECTIONS = {
+            MessageProvider.MessageColumns._ID,
+            MessageProvider.MessageColumns.SENDER,
+            MessageProvider.MessageColumns.SEND_DATE,
+            MessageProvider.MessageColumns.SUBJECT,
+            MessageProvider.MessageColumns.PREVIEW,
+            MessageProvider.MessageColumns.UNREAD,
+            MessageProvider.MessageColumns.HAS_ATTACHMENTS,
+            MessageProvider.MessageColumns.URI
+    };
+
+    private Context mContext;
+    private ArrayList<MailItem> mailItems;
+    private int count;
+
+    public MailListRemoteViewFactory(Context context, Intent intent) {
+        mContext = context;
+    }
+
+    @Override
+    public void onCreate() {
+        mailItems = new ArrayList<>(25);
+    }
+
+    @Override
+    public void onDataSetChanged() {
+        final long identityToken = Binder.clearCallingIdentity();
+        mailItems = new ArrayList<>(25);
+        Cursor cursor = mContext.getContentResolver().query(
+                MessageProvider.CONTENT_URI.buildUpon().appendPath("inbox_messages").build(),
+                MAIL_LIST_PROJECTIONS,
+                null,
+                null,
+                MessageProvider.MessageColumns.SEND_DATE + "DESC");
+        while (cursor.moveToNext()) {
+            mailItems.add(new MailItem(
+                        cursor.getString(0), /* id */
```

Comment:
My concern here was that i would create 8 local variable at every iteration thereby adding a (slight albeit) memory overhead. That's why i added the comments.